<a href="https://colab.research.google.com/github/M2kishore/Machine-Learning/blob/main/cb_en_u4cse19127_Eval2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [139]:
from scipy.io import arff
import pandas as pd
import numpy as np

In [140]:
df = pd.read_csv('/content/Chronic_Kidney_Disease.csv')

In [141]:
df.isna().sum()

id         0
age        9
bp        12
sg        47
al        46
su        49
rbc      152
pc        65
pcc        4
ba         4
bgr       44
bu        19
sc        17
sod       87
pot       88
hemo      52
pcv       70
wc       105
rc       130
htn        2
dm         2
cad        2
appet      1
pe         1
ane        1
class      0
dtype: int64

In [142]:
df = df.dropna()

In [143]:
df.describe()

id         age          bp  ...         sod         pot        hemo
count  158.000000  158.000000  158.000000  ...  158.000000  158.000000  158.000000
mean   274.841772   49.563291   74.050633  ...  138.848101    4.636709   13.687342
std    102.998517   15.512244   11.175381  ...    7.489421    3.476351    2.882204
min      3.000000    6.000000   50.000000  ...  111.000000    2.500000    3.100000
25%    243.000000   39.250000   60.000000  ...  135.000000    3.700000   12.600000
50%    298.500000   50.500000   80.000000  ...  139.000000    4.500000   14.250000
75%    355.750000   60.000000   80.000000  ...  144.000000    4.900000   15.775000
max    399.000000   83.000000  110.000000  ...  150.000000   47.000000   17.800000

[8 rows x 12 columns]

In [144]:
df = df.drop("id", axis=1)

In [145]:
df = pd.get_dummies(df)
df

age    bp     sg   al  ...  ane_no  ane_yes  class_ckd  class_notckd
3    48.0  70.0  1.005  4.0  ...       0        1          1             0
9    53.0  90.0  1.020  2.0  ...       0        1          1             0
11   63.0  70.0  1.010  3.0  ...       1        0          1             0
14   68.0  80.0  1.010  3.0  ...       1        0          1             0
20   61.0  80.0  1.015  2.0  ...       0        1          1             0
..    ...   ...    ...  ...  ...     ...      ...        ...           ...
395  55.0  80.0  1.020  0.0  ...       1        0          0             1
396  42.0  70.0  1.025  0.0  ...       1        0          0             1
397  12.0  80.0  1.020  0.0  ...       1        0          0             1
398  17.0  60.0  1.025  0.0  ...       1        0          0             1
399  58.0  80.0  1.025  0.0  ...       1        0          0             1

[158 rows x 179 columns]

In [146]:
df = df.drop('class_notckd',axis=1)

In [147]:
#from sklearn import preprocessing
#df = preprocessing.normalize(df)
#df

**2**<br>
i. Model 1 – Logistic Regression

ii. Model 2 – Decision Tree (ID3)

iii. Model 3 – Random Forest

iv. Model 4 – K Nearest Neighbour

In [169]:
from sklearn.model_selection import train_test_split
X = df.iloc[:,:-1].values
Y = df.iloc[:,-1].values.reshape(-1,1)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 42)

In [201]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score,f1_score,recall_score
clf = DecisionTreeClassifier()
DTclf = clf.fit(X,Y)
prediction = DTclf.predict(X_test)
accuracy_score(y_test,prediction)
scores = []
for i in range(1,6):
  model = DecisionTreeClassifier(max_depth=i)
  model.fit(X_train,y_train)
  y_pred = model.predict(X_test)
  score = accuracy_score(y_test,y_pred)
  scores.append(score)
print(scores)
print(sum(scores)/len(scores))

[0.96875, 0.96875, 1.0, 0.96875, 1.0]
0.98125


In [202]:
print(f1_score(y_test,y_pred))
print(recall_score(y_test,y_pred))

1.0
1.0


In [173]:
from sklearn.metrics import confusion_matrix
cm= confusion_matrix(y_test, y_pred)
cm 

array([[23,  0],
       [ 1,  8]])

In [184]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators=5)
model.fit(X_train,y_train)
y_pred = model.predict(X_test)
accuracy_score(y_test,y_pred)
scores = []
for i in range(1,6):
  model = RandomForestClassifier(n_estimators=i)
  model.fit(X_train,y_train)
  y_pred = model.predict(X_test)
  score = accuracy_score(y_test,y_pred)
  scores.append(score)
print(scores)
print(sum(scores)/len(scores))

[0.96875, 0.96875, 1.0, 1.0, 1.0]
0.9875


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  if __name__ == '__main__':
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  if __name__ == '__main__':
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example

In [156]:
from sklearn.metrics import confusion_matrix
cm= confusion_matrix(y_test, y_pred)
cm 

array([[23,  0],
       [ 2,  7]])

In [180]:
from sklearn.neighbors import KNeighborsClassifier  
classifier= KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2 )  
classifier.fit(X_train, y_train)
y_pred= classifier.predict(X_test)
accuracy_score(y_test,y_pred)
scores = []
for i in range(1,6):
  model = KNeighborsClassifier(n_neighbors=i, metric='minkowski', p=2 )
  model.fit(X_train,y_train)
  y_pred = model.predict(X_test)
  score = accuracy_score(y_test,y_pred)
  scores.append(score)
print(scores)
print(sum(scores)/len(scores))

[0.90625, 0.9375, 0.9375, 0.90625, 0.90625]
0.91875


/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

In [161]:
from sklearn.metrics import confusion_matrix
cm= confusion_matrix(y_test, y_pred)
cm 

array([[23,  0],
       [ 3,  6]])

In [181]:
from sklearn.linear_model import LogisticRegression  
classifier= LogisticRegression(random_state=42)  
classifier.fit(X_train, y_train)  
y_pred= classifier.predict(X_test)
accuracy_score(y_test,y_pred)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:985: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


0.96875

In [168]:
from sklearn.metrics import confusion_matrix
cm= confusion_matrix(y_test, y_pred)
cm 

array([[23,  0],
       [ 1,  8]])

random forest = 98.125%<br>
ID3 = 98.75%<br>
Logistic regression = 96.87%<br>
KNN = 90.625%<br>


* This accuracy is attained by dropping a lot of nan values as false positives 
and false negatives needs to be very low and the model presision needs to be very accurate.<br>
* Tree based models are performing better than linear models like Logistic regression<br>
* KNN is not so effective as there is no training involved and the decisions are blindly taken on the basis of the nearedt element
* There are little bit fluctuations when running the code again as there is some learning happening each and everytime the model is fit




In [155]:
from sklearn.neighbors import KNeighborsClassifier  
classifier= KNeighborsClassifier(n_neighbors=11, metric='minkowski', p=2 )  
classifier.fit(X_train, y_train)
y_pred= classifier.predict(X_test)
accuracy_score(y_test,y_pred)


/usr/local/lib/python3.7/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.9375

The accuracy increases with the number of neighbours sqrt(143) is optimal 11 was taken which gave 93.75% accuracy

Final thoughts<br>
By considering the above examples Tree based models give better accuracy than other models for this Dataset which is categorical.